flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/taskmanager/TaskManagerLogsHandler.java

```diff
@@ -0,0 +1,98 @@
+import org.apache.flink.runtime.rest.messages.taskmanager.TaskManagerIdPathParameter;
+import org.apache.flink.runtime.rest.messages.taskmanager.TaskManagerMessageParameters;
+import org.apache.flink.runtime.webmonitor.RestfulGateway;
+import org.apache.flink.runtime.webmonitor.retriever.GatewayRetriever;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpResponseStatus;
+
+import javax.annotation.Nonnull;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.stream.Collectors;
+
+/**
+ * Handler which serves detailed TaskManager log list information.
+ */
+public class TaskManagerLogsHandler extends AbstractTaskManagerHandler<RestfulGateway, EmptyRequestBody, LogsInfo, TaskManagerMessageParameters> {
+
+	private final GatewayRetriever<ResourceManagerGateway> resourceManagerGatewayRetriever;
+
+	public TaskManagerLogsHandler(
+			GatewayRetriever<? extends RestfulGateway> leaderRetriever,
+			Time timeout,
+			Map<String, String> responseHeaders,
+			MessageHeaders<EmptyRequestBody, LogsInfo, TaskManagerMessageParameters> messageHeaders,
+			GatewayRetriever<ResourceManagerGateway> resourceManagerGatewayRetriever) {
+		super(leaderRetriever, timeout, responseHeaders, messageHeaders, resourceManagerGatewayRetriever);
+
+		this.resourceManagerGatewayRetriever = Preconditions.checkNotNull(resourceManagerGatewayRetriever);
+	}
+
+	@Override
+	protected CompletableFuture<LogsInfo> handleRequest(
+			@Nonnull HandlerRequest<EmptyRequestBody, TaskManagerMessageParameters> request,
+			@Nonnull ResourceManagerGateway gateway) throws RestHandlerException {
+		final ResourceID taskManagerId = request.getPathParameter(TaskManagerIdPathParameter.class);
+		final ResourceManagerGateway resourceManagerGateway = getResourceManagerGateway(resourceManagerGatewayRetriever);
+		final CompletableFuture<Collection<Tuple2<String, Long>>> logsWithLengthFuture = resourceManagerGateway.requestTaskManagerLogList(taskManagerId, timeout);
+
+		return logsWithLengthFuture.thenApply(logName2Sizes -> {
+			if (null != logName2Sizes) {
+				Collection<LogInfo> logs = logName2Sizes.stream().map(logName2Size -> new LogInfo(logName2Size.f0, logName2Size.f1)).collect(Collectors.toSet());
+				return new LogsInfo(logs);
+			} else {
+				return LogsInfo.empty();
+			}
+		}).exceptionally(
+				(Throwable throwable) -> {
+					final Throwable strippedThrowable = ExceptionUtils.stripExecutionException(throwable);
+
+					if (strippedThrowable instanceof UnknownTaskExecutorException) {
+						throw new CompletionException(
+							new RestHandlerException(
+								"Could not find TaskExecutor " + taskManagerId + '.',
+								HttpResponseStatus.NOT_FOUND,
+								strippedThrowable));
+					} else {
+						throw new CompletionException(strippedThrowable);
```

Comment:
It doesn't seem right that we return internal server error if the file does not exist.

Reply:
What about 204 which means not content
```java
new RestHandlerException("Could not find any file.", HttpResponseStatus.NO_CONTENT,strippedThrowable));
```

Reply:
Why not 404? Requesting an invalid file should be a client error, i.e., 4xx.